### DISCUSSION TITLE
Facets for navigation feeds ?

### DISCUSSION THREAD

**Discussion by gotson** [opds-community/drafts#69] [Ideas]
While trying to implement OPDS 2 i lodged https://github.com/edrlab/thorium-reader/issues/1980 when facets would not be displayed in Thorium. Turns out they are only supported for publication feeds.

In [Komga](https://komga.org/), and i'm sure others, navigation is done using subsections, for example a Library that will display series. Series will be another container, so will be rendered in a navigation feed. When exploring the series, it will list publications (books).

Using facets for filtering a list of series makes sense, for example to filter by publisher, language etc.

However facets cannot be used for that.

Is that something that could be changed ?

Else a workaround is to use traditional navigation links to perform filtering. Using facets would provide better context to the OPDS client, for example by using a filter panel in a different place than traditional navigation links.

**Comment by gotson** [comment]
So i have implemented a workaround using navigation links, however since those are not understood as facets in Thorium (could be another reader), it displays some breadcrumbs that keeps piling up.

<img width="1396" alt="image" src="https://github.com/opds-community/drafts/assets/2139133/7093f16c-32f6-431e-bd42-f3a66d2df85c">


After clicking some publishers a few times:

<img width="1145" alt="image" src="https://github.com/opds-community/drafts/assets/2139133/0d31a107-b5bc-46a6-bad9-b94254770c43">

I have not yet got facets working in Thorium, but i think fundamentally it would be handled differently in the client compared to navigation links.

**Comment by HadrienGardeur** [comment]
Facets have historically been designed to obtain a subset or change the order of an acquisition/publication feed (they were added in the OPDS 1.1 revision of the spec).

Most OPDS catalogs and clients only use navigation feeds and groups sporadically, for example for providing a main menu on a homepage or for listing categories. 

To explore publications by publisher/series/language, most catalogs rely on a contextual approach rather than a hierarchical one. In OPDS 2.0, you can include links to your metadata and in Aldiko for example, we'll fetch some of those links to display carousels on a publication's page. Even if a carousel is not displayed, this allows the user to tap/click on the publisher's name to see everything available for them.

From a spec perspective, catalogs are free to include facets on navigation feeds or mixed feeds that contain groups or a mix of publications/navigation. But since facets were not really designed with that in mind, they tend to be ignored by clients. We could update the spec to reflect that and have an open conversation with a number of clients to nudge them towards wider support for facets.

**Reply by gotson** [reply]
Seems facets are not allowed on navigation links from a spec perspective though https://github.com/edrlab/thorium-reader/issues/1980#issuecomment-1689467000

> To explore publications by publisher/series/language, most catalogs rely on a contextual approach rather than a hierarchical one. In OPDS 2.0, you can include links to your metadata and in Aldiko for example, we'll fetch some of those links to display carousels on a publication's page. Even if a carousel is not displayed, this allows the user to tap/click on the publisher's name to see everything available for them.

But that means you would already be in the publication already. If you want to slice and dice a library, facets should be the way to go.

**Reply by HadrienGardeur** [reply]
Well, they're technically allowed since they wouldn't get rejected by the JSON Schema but we clearly point towards use cases for feeds containing strictly publications.

**Comment by HadrienGardeur** [comment]
This item is a good candidate for a future discussion in an OPDS monthly call. 

I'm open to the idea of having facets on navigation feeds in OPDS 2.0 but I wonder how others feel about this.

**Comment by ajslater** [comment]
Hello from the future. Facets on navigation feeds seem like a good idea for clients to provide visual cues that more of a slice is being taken than a group navigation.

However from my experience with OPDS v1.0 clients, implementation uptake among clients will be scattershot and server devs would have to provide navigation links as fallbacks. But when there is a nice UI for it in a client, I do like to use it.

**Reply by HadrienGardeur** [reply]
We're currently working on Thorium Mobile for iOS with a target release planned in April.

Since our OPDS support will primarily target OPDS 2.0 features (OPDS 1.x will be handled at a parser-level but in a way where we don't care about XML vs JSON), we'll make sure to support and display facets in all feeds.
We'll also use this opportunity to rework support for OPDS in the Readium Swift toolkits, which will hopefully help with adoption in other clients.

Unlike the screenshot posted above, we have zero plans for breadcrumbs, they're incompatible with complex catalogs that rely on facets and contextual navigation (links at a publication level).
